### PR TITLE
Implemented slope reconstruction for 1-phase boundaries and JunctionOneToOne1Phase

### LINKED ISSUE #20532: Implement slope reconstruction for boundaries and junctions in THM
## Reason
Solution artifacts can appear at boundaries and around junctions because slope reconstructions is implemented everywhere else.

## Design
This would implement slope reconstruction for boundaries and junctions.

## Impact
This will change the solution behavior when slope reconstruction is used with boundaries and junctions.

### PR DESCRIPTION
**Pull request by joshuahansel**
These changes implement slope reconstruction for all 1-phase boundaries and `JunctionOneToOne1Phase`.

This was tested on a new pressure drop problem. This problem was provided with a spatial discretization study and tested before and after the change. The results are shown in the following plot:
pressure_drop
The important observations are:

- Minmod, MC, and Superbee appear to perform identically for this problem (before and after), so that’s why you can only see Superbee, the last of them plotted.
- Before the changes, “Full” has visible artifacts at junctions at boundaries (slight in this problem), and after the changes, it does not.
- Performing slope reconstruction at the boundaries makes a huge difference for all slope reconstruction options.

Additionally, these changes were tested on the following input file: 
[02_core.txt](https://github.com/idaholab/moose/files/9112866/02_core.txt)
The before and after are shown in the following plot:
Screen Shot 2022-07-13 at 11 09 50 AM
Slope reconstruction does not completely resolve all artifacts, but it does significantly reduce them.

Closes #20532

### CODE REVIEW

File: modules/thermal_hydraulics/src/userobjects/ADJunctionOneToOne1PhaseUserObject.C [hunk 1]
```diff
@@ -128,41 +173,95 @@ ADJunctionOneToOne1PhaseUserObject::threadJoin(const UserObject & uo)
...
+  if (!THM::areParallelVectors(_directions[0], _directions[1]))
+    mooseError(_junction_name, ": The connected channels must be parallel at the junction.");
+
+  const auto & W0_avg = _primitive_solutions[0];
+  const auto & W1_avg = _primitive_solutions[1];
 
-  _fluxes[0] = _numerical_flux.getFlux(
-      _local_side_ids[0], _elem_ids[0], true, _solutions[0], _solutions[1], _normal[0]);
+  // Multiplier for possibly different coordinate systems
+  const Real dir_mult = _directions[0] * _directions[1];
```
Comment (licharlot) [inline modules/thermal_hydraulics/src/userobjects/ADJunctionOneToOne1PhaseUserObject.C:219]: Is it necessary with the assumption that the channels are parallel?

File: modules/thermal_hydraulics/test/tests/problems/pressure_drop/pressure_drop.i [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+      molar_mass = 2.016e-3
+      k = 0.437
+      mu = 3e-5
+    []
+  []
+[]
+
+[Closures]
+  [no_closures]
+    type = Closures1PhaseNone
```
Comment (licharlot) [inline modules/thermal_hydraulics/test/tests/problems/pressure_drop/pressure_drop.i:34]: Is there a reason not to use the simple closures? the friction factor is constant.

File: modules/thermal_hydraulics/test/tests/problems/pressure_drop/gold/pressure_drop_csv_pressure_vpp_FINAL.csv [hunk 3]
```diff
@@ -0,0 +1,101 @@
+id,p,x,y,z
+0.005,7000000.2679436,0.005,0,3.0616169978684e-19
```
Comment (licharlot) [inline modules/thermal_hydraulics/test/tests/problems/pressure_drop/gold/pressure_drop_csv_pressure_vpp_FINAL.csv:2]: Could you increase the mass flow rate or friction factor to get a large pressure drop?

**Review by licharlot — CHANGES_REQUESTED** [review]

File: modules/thermal_hydraulics/src/userobjects/ADJunctionOneToOne1PhaseUserObject.C (hunk 1 above)
Comment (joshuahansel) [reply modules/thermal_hydraulics/src/userobjects/ADJunctionOneToOne1PhaseUserObject.C:219]: The channels are parallel but may be in opposite directions, so `dir_mult` is equal to +/-1.

File: modules/thermal_hydraulics/test/tests/problems/pressure_drop/pressure_drop.i (hunk 2 above)
Comment (joshuahansel) [reply modules/thermal_hydraulics/test/tests/problems/pressure_drop/pressure_drop.i:34]: Nope, you were right. I don't know why I did that.

File: modules/thermal_hydraulics/test/tests/problems/pressure_drop/gold/pressure_drop_csv_pressure_vpp_FINAL.csv (hunk 3 above)
Comment (joshuahansel) [reply modules/thermal_hydraulics/test/tests/problems/pressure_drop/gold/pressure_drop_csv_pressure_vpp_FINAL.csv]: What happened is that in the test I only took 5 very small time steps, so it didn't have time to get close to the end pressure drop. It turned out that I could increase time step size significantly, so this makes the pressure drops much bigger.

File: modules/thermal_hydraulics/src/userobjects/ADJunctionOneToOne1PhaseUserObject.C (hunk 1 above)
Comment (licharlot) [reply modules/thermal_hydraulics/src/userobjects/ADJunctionOneToOne1PhaseUserObject.C:219]: Ok, I got confused, this cannot do 180 degree flow turn, but the direction of the flow channels could be in opposite directions.

File: modules/thermal_hydraulics/test/tests/problems/pressure_drop/gold/pressure_drop_csv_pressure_vpp_FINAL.csv (hunk 3 above)
Comment (licharlot) [reply modules/thermal_hydraulics/test/tests/problems/pressure_drop/gold/pressure_drop_csv_pressure_vpp_FINAL.csv]: Ok it looks better know. Making sure we are not capturing noise.

**Review by licharlot — APPROVED** [review]

**Review by loganharbour — APPROVED** [review]
Approving merge with the python failure; said failure is just a network issue

**Pull request merged**